docs/design-documents/platform/clocks/clocks.md

```diff
@@ -0,0 +1,418 @@
+```
+
+Namespace directives should only be used in source files to avoid unexpected name collision for users. Header files intended for general use will need to fully qualify their use of Chrono with `std::chrono` prefixes. This also precludes use of literals:
+
+```C++
+// In a header file
+#include <chrono>
+void my_function(std::chrono::microseconds timeout = std::chrono::seconds{1});
+```
+
+### Using the RTOS clock
+
+**Basic delays**
+
+```C++
+for (;;) {
+    LED1 = !LED1;
+    ThisThread::sleep_for(500ms);
+}
+```
+
+**Prefer absolute time**
+
+Using absolute time aids precision, guaranteeing that there is no drift due to code execution time. This applies to all clocks.
+
+```C++
+auto next_wake = Kernel::Clock::now();
+for (;;) {
+    LED1 = !LED1;
+    next_wake += 500ms;
+    ThisThread::sleep_until(next_wake);
+}
+```
+
+Periodic callback APIs such as `Ticker` or `EventQueue::call_every` use this mechanism. If manually rescheduling events yourself, you should use the same technique, rather than using `call_in` within a callback.
+
+**Measuring elapsed time via the RTOS**
+
+```C++
+auto start = Kernel::Clock::now(); // type will be Kernel::Clock::time_point
+do_operation();
+milliseconds elapsed_time = Kernel::Clock::now() - start;
+printf("elapsed time = %d ms\n", int(elapsed_time.count()));
+```
+
+Explicitly stating that `elapsed_time` is `milliseconds` provides a cross-check that the `count()` for the print really is milliseconds. If the clock had lower resolution, it would be implicitly converted to milliseconds. If the clock had higher resolution, it would be a compilation error - see "being more generic" below to handle this.
+
+### Using the high-resolution clock
+
+**Measuring elapsed time using Timer**
+```C++
+template<class F>
+microseconds time_operation(const F &operation)
+{
+    Timer timer;
+    timer.start(); // deep sleep is automatically locked while timer is running
+    operation();
+    return timer.elapsed_time();
+} // Deep sleep automatically unlocked when timer is destroyed.
+
+microseconds print_time = time_operation([] { printf("Hello!\n"); });
+printf("Printing that took %d us\n", int(print_time.count()));
+```
+
+This example relies on the destruction of the timer to unlock the clock. If using a timer that doesn't get destroyed, you must remember to stop it manually to allow deep sleep to be entered again.
+
+**Recording high-resolution time points**
+```C++
+HighResClock::lock(); // take lock to ensure clock is steady (by preventing deep sleep)
+auto t1 = HighResClock::now();
+do_operation1();
+auto t2 = HighResClock::now();
+do_operation2();
+auto t3 = HighResClock::now();
+HighResClock::unlock();
+microseconds total = t3 - t1;
+microseconds stage1 = t2 - t1;
+microseconds stage2 = t3 - t2; 
+printf("total time   = %d us\n", int(total.count()));
+printf("stage 1 time = %d us\n", int(stage1.count()));
+printf("stage 2 time = %d us\n", int(stage2.count()));
+```
+
+**Blinking an LED at 1kHz**
```

Comment:
some example have a line break before the code block, some don't.

when converting to commonmark with pandoc, a new line is added after `...**`.